### DISCUSSION TITLE
E220-400M22S (LLCC68) RXEN TXEN

### DISCUSSION THREAD

**Discussion by thanosbi** [jgromes/RadioLib#1320] [Q&A] [answered]
E220-400M22S has RXEN TXEN these pins. Can someone explain what is their role? 
Ebyte info is very little, says for RXEN: "RF switch receive control pin, connect external MCU IO, high level effective"
and "RF switch emission control pin, connected to external MCU IO or dio2, high level effective" for the TXEN.

**Comment by HeadBoffin** [comment]
Some similar discussion here: https://github.com/jgromes/RadioLib/discussions/1265

Found with a search of just RXEN with many other hits, so there may be some other info to be looked at.

**Comment by S5NC** [comment, marked as answer]
_Explanation highly adapted from from https://github.com/meshtastic/firmware/blob/master/variants/EBYTE_ESP32-S3/variant.h (I wrote the text about the options in that file)_

The module connects the exposed TXEN and RXEN connections to an RF switch which connects the antenna to the transmission output of the SX1262 (the LoRa IC in the module) or the receiving input.

Some modules also have Low Noise Amplifiers (LNA) and Power Amplifiers (PA) but this module doesn't have this.

In your case, the antenna can only be connected to the transmission output of the SX1262 or its receiving input. Maybe the RF switch disconnects the antenna completely when RXEN and TXEN are the same, but I don't know this as I don't have the datasheet for the RF switch used.

The SX126x's function "DIO2 as RF Switch" sets DIO2 to high when it is about to transmit. You can configure the delay between setting DIO2 high and starting the transmission.
If you don't use this function, DIO2 can be used as an interrupt pin (in addition to DIO1), however one interrupt pin is generally enough.

Your module's datasheet says when transmitting, have TXEN HIGH and RXEN LOW. When receiving, have TXEN LOW and RXEN HIGH.

You can do this multiple ways:
1. (Commonly used, used to have more manual control over the switching process or to maintain access to DIO2 to use as an additional interrupt pin) Connect TXEN to an MCU pin. Connect RXEN to an MCU pin. Let RadioLib control TXEN and RXEN manually.
2. (Commonly used to save 2 MCU pins when DIO2 is not needed as an interrupt) Connect DIO2 to TXEN. Invert DIO2 using hardware and connect that inverted DIO2 signal to RXEN. Use the function of the SX1262 called DIO2 as RF Switch which you can access via RadioLib.
3. (Commonly used, used to save one MCU pin when DIO2 is not needed as an interrupt) Connect DIO2 to TXEN. Use DIO2 as RF Switch which you can access via RadioLib. Connect RXEN to an MCU pin. Let RadioLib control RXEN manually.
4. (Crazy but you could do it if you really wanted to...) Connect TXEN to an MCU pin. Let RadioLib control TXEN manually. Invert DIO2 using hardware and connect that inverted DIO2 signal to RXEN. Use DIO2 as RF Switch which you can access via RadioLib.

Use option 1 if you can use 2 MCU pins for RF switching.
Use option 2 if you can use 1 MCU pin for RF switching and don't need DIO2 as an interrupt.
Use option 3 if you can use no MCU pins for RF switching and can add a hardware logical inversion and don't need DIO2 as an interrupt.
Use option 4 if you can use 1 MCU pin for RF switching and can add a hardware logical inversion and don't need DIO2 as an interrupt. You should never have to use this option, because you could always use option 2 instead and avoid adding extra hardware.

For your module you don't need this, but for information: If you ever need TXEN and RXEN to be both off (or both on) always use option 1. For example, for modules with PAs and LNAs you might want to be able to do this (when you are neither TXing or RXing to save power if the PA is enabled by TXEN and the LNA is enabled by TXEN).

**Personally for your module I would use option 1 for the most flexibility, option 2 if I am slightly short on MCU pins, and option 3 if I am even more short on MCU pins are can add a hardware inverter, or if I am designing the PCB and want to be elegant or be sure I avoid undefined behaviour cases with an RF switch controlled by TXEN and RXEN**

**Reply by thanosbi** [reply]
I think you are absolutely right! Already tried "option 1" seems to work ok (almost for a week now) need to check it  at long distance. 
The on board RFswitch(?) is marked 404 2220 in sot23-6 package searched but cant find anything. Generally these modules have weird behavior. Their chips supposed to be llcc68 but they are reporting as sx1261 [https://github.com/jgromes/RadioLib/issues/1329](url)

**Reply by djObsidian** [reply]
I asked ebyte support about that, their reply was
> PE4259 has two kinds of control logic, we use 01 and 10 control logic, I believe you can see the TXEN and RXEN pins describe the control on the corresponding manual, please follow the instructions on the manual to test.

It seems, like they are using PE4259, so it's possible to use single signal control mode of this RF swtich. Just pull RXEN high and by controlling TXEN you can switch TX and RX. SWL2001 uses option 3 in their examples, but the firmware doesn't controls RXEN (called ANT SW there), it is just pulled up at init and that's it.
https://github.com/Lora-net/SWL2001/blob/master/lbm_examples/smtc_hal_l0_LL/smtc_hal_mcu.c#L313
Tested this with E220-900M22S, works fine. 
LLCC68 is a simplified version of SX1262, probably even binned, no wonder it is defined differently.

**Reply by S5NC** [reply]
> It seems, like they are using PE4259, so it's possible to use single signal control mode of this RF swtich. Just pull RXEN high and by controlling TXEN you can switch TX and RX. SWL2001 uses option 3 in their examples, but the firmware doesn't controls RXEN (called ANT SW there), it is just pulled up at init and that's it. https://github.com/Lora-net/SWL2001/blob/master/lbm_examples/smtc_hal_l0_LL/smtc_hal_mcu.c#L313 Tested this with E220-900M22S, works fine. LLCC68 is a simplified version of SX1262, probably even binned, no wonder it is defined differently.

![image](https://github.com/user-attachments/assets/b3c16786-cf57-48e3-b63f-ca7ee1f184a2)

In the [PE4259 datasheet](https://www.psemi.com/pdf/datasheets/pe4259ds.pdf) it says

> Control Logic Input
> The PE4259 is a versatile RF CMOS switch that supports two operating control modes; single-pin control mode and complementary-pin control mode.
>
> Single-pin control mode enables the switch to operate with a single control pin (pin 4) supporting a +3-volt CMOS logic input, and requires a dedicated +3-volt power supply connection on pin 6 (VDD). This mode of operation reduces the number of control lines required and simplifies the switch control interface typically derived from a CMOS µProcessor I/O port.
>
> Complementary-pin control mode allows the switch to operate using complementary control pins CTRL and CTRL (pins 4 and 6), that can be directly driven by +3-volt CMOS logic or a suitable µProcessor I/O port. **This enables the PE4259 to be used as a potential alternate source for SPDT RF switch products used in positive control voltage mode and operating within the PE4259 operating limits.**

The emboldened sentence may suggest there is better isolation when using the complementary mode. pSemi should make this clearer in their datasheet or add an application note, I don't see any relevant ones about the performance between modes. (If the switches aren't discrete, or the connections aren't **only** _RFC to RF1_ or _RFC to RF2_ at a time this could be a reason for a difference.)

You could measure the isolation between modes if you like or ask Ebyte if this is why they put complementary mode in their M22S model (for the boards with a PA, like the M30S and M33S, CTRL would need to be used to control the power supply to the PA because it is the one that changes in single-pin control mode, so single-pin control mode is not ruled out as the LNA isn't switched on/off in their modules).
